src/stan/io/reader.hpp

```diff
@@ -299,15 +312,92 @@ class reader {
    * for more information.
    *
-   * @param m Number of rows.
-   * @param n Number of columns.
+   * @param n Number of rows.
+   * @param m Number of columns.
    * lp Log probability to increment.
    * @return Matrix made up of the next scalars.
    */
-  inline matrix_t matrix_constrain(size_t m, size_t n, T & /*lp*/) {
-    if (m == 0 || n == 0)
-      return matrix_t(m, n);
-    return map_matrix_t(&scalar_ptr_increment(m * n), m, n);
+  inline matrix_t matrix_constrain(size_t n, size_t m, T& /*lp*/) {
+    return matrix(n, m);
+  }
+
+  /**
+   * Return a sparse matrix of the specified dimensionality and number of
+   * nonzero elements made up of the next scalars arranged in
+   * column-major order.
+   *
+   * @tparam VecR Vector with an integral type
+   * @tparam VecC Vector with an integral type
+   * @param vec_r rows corresponding the nonzero entries.
+   * @param vec_c columns corresponding the nonzero entries.
+   * @param n Number of rows.
+   * @param m Number of columns.
+   * @return Eigen::SparseMatrix made up of the next scalars.
+   */
+  template <typename VecR, typename VecC,
+            typename = require_all_std_vector_vt<std::is_integral, VecR, VecC>>
+  inline Eigen::SparseMatrix<T> sparse_matrix(const VecR& vec_r,
+                                              const VecC& vec_c, size_t n,
+                                              size_t m) {
+    stan::math::check_matching_sizes("sparse_matrix", "vec_r", vec_r, "vec_c",
+                                     vec_c);
+    Eigen::SparseMatrix<T> ret_mat(n, m);
+    if (m == 0 || n == 0) {
+      return ret_mat;
+    }
+    using triplet_type = Eigen::Triplet<T>;
+    std::vector<triplet_type> triplet_list(vec_r.size());
+    for (auto i = 0; i < vec_r.size(); i++) {
+      triplet_list.emplace_back(
+          triplet_type(vec_r[i], vec_c[i], scalar_ptr_increment(1)));
+    }
+    ret_mat.setFromTriplets(triplet_list.begin(), triplet_list.end());
+    return ret_mat;
+  }
+
+  /**
+   * Return a sparse matrix of the specified dimensionality and number of
+   * nonzero elements made up of the next scalars arranged in
+   * column-major order.
+   *
+   * @tparam VecR Vector with an integral type
+   * @tparam VecC Vector with an integral type
+   * @param vec_r rows corresponding the nonzero entries.
+   * @param vec_c columns corresponding the nonzero entries.
+   * @param n Number of rows.
+   * @param m Number of columns.
+   * @return Eigen::Matrix made up of the next scalars.
+   */
+  template <typename VecR, typename VecC,
+            typename = require_all_std_vector_vt<std::is_integral, VecR, VecC>>
+  inline Eigen::SparseMatrix<T> sparse_matrix_constrain(const VecR& vec_r,
+                                                        const VecC& vec_c,
+                                                        size_t n, size_t m) {
+    return sparse_matrix(std::forward<VecR>(vec_r), std::forward<VecC>(vec_c),
```

Comment:
[question]
Since they arguments are only constant references coming in, what does forward do here?